homeassistant/components/mqtt/client.py

```diff
@@ -435,12 +435,13 @@ def no_more_acks() -> bool:
             """Return False if there are unprocessed ACKs."""
             return not bool(self._pending_operations)
 
-        # wait for ACK-s to be processesed (unsubscribe only)
+        # wait for ACK-s to be processesed
         async with self._pending_operations_condition:
             await self._pending_operations_condition.wait_for(no_more_acks)
 
         # stop the MQTT loop
-        await self.hass.async_add_executor_job(stop)
+        async with self._paho_lock:
+            await self.hass.async_add_executor_job(stop)
```

Comment:
Since this is a MQTT client operations this should be handled with the Paho Lock